### PR TITLE
Moved max_read_level inside of db

### PR DESCRIPTION
**Pull request by Shivs11**
## What changed?
This PR moves `max_read_level` inside of db so that `db` becomes the single source of truth for this variable. 

## Why?
While working on #5593, we realized there were concurrency issues, such as deadlock, which were spinning up when `db` tried to read `max_read_level` for resetting the `approximateBacklogCounter`. Thus, moving `max_read_level` inside of db will solve that issue and shall also serve towards the purpose of making `db` the single source of truth for most shared resources in the near future.

## How did you test it?
Current unit tests.

## Potential risks
None

## Is hotfix candidate?
No

### CODE REVIEW

File: service/matching/db.go [hunk 1]
```diff
@@ -188,11 +204,24 @@ func (db *taskQueueDB) UpdateState(
...
-	return db.store.CreateTasks(
+
+	maxReadLevel := int64(0)
+	var tasks []*persistencespb.AllocatedTaskInfo
+	for i, req := range reqs {
+		tasks = append(tasks, &persistencespb.AllocatedTaskInfo{
+			TaskId: taskIDs[i],
+			Data:   req.taskInfo,
+		})
+		maxReadLevel = taskIDs[i]
```
Comment (tdeebswihart) [inline service/matching/db.go:220]: Is this correct? Are tasks guaranteed to be in order of increasing ID?

File: service/matching/task_writer.go [hunk 2]
```diff
@@ -227,29 +222,15 @@ writerLoop:
...
-					TaskId: taskIDs[i],
-					Data:   req.taskInfo,
-				})
-				maxReadLevel = taskIDs[i]
-			}
-
-			resp, err := w.appendTasks(ctx, tasks)
+			resp, err := w.appendTasks(ctx, taskIDs, reqs)
 			// Update the maxReadLevel after the writes are completed, but before we send the response,
 			// so that taskReader is guaranteed to see the new read level when SpoolTask wakes it up.
```
Comment (ShahabT) [inline service/matching/task_writer.go:233]: please move the comment to the new place (and update is if needed).

File: service/matching/task_writer.go [hunk 3]
```diff
@@ -227,29 +222,15 @@ writerLoop:
...
-			if maxReadLevel > 0 {
```
Comment (ShahabT) [inline service/matching/task_writer.go:250]: is this check dropped by intention?

File: service/matching/db.go [hunk 4]
```diff
@@ -188,11 +204,24 @@ func (db *taskQueueDB) UpdateState(
...
+	}
+
+	db.maxReadLevel = maxReadLevel
```
Comment (ShahabT) [inline service/matching/db.go:223]: in old implementation update is conditional on `maxReadLevel > 0` to be safe we need to keep that check or something similar in here.
Also, we're changing order of db.maxReadLevel assignment and bringing it to before CreateTask. Not sure it's desired.

File: service/matching/db.go (hunk 1 above)
Comment (ShahabT) [reply service/matching/db.go:220]: yup, this part is just moved from task_writer, same logic is kept.

Comment (Shivs11) [reply service/matching/db.go:220]: Yep, just moved into a different place

File: service/matching/task_writer.go (hunk 3 above)
Comment (Shivs11) [reply service/matching/task_writer.go:250]: seemed to have missed the fact that if we don't check it, as being done here, it might set it to 0 and could break things. Shall add the case where it updates inside of `db`.

My bad!

File: service/matching/db.go (hunk 4 above)
Comment (Shivs11) [reply service/matching/db.go:223]: I had thought of this during it's implementation and would like to raise the question:
Is it possible, that we add tasks, and the check `maxReadLevel > 0` fails? 

After spending some time thinking about this, I found out that the answer to that is when a request would have 0 tasks for addition. In that case, `maxReadLevel` would remain 0 since we do not iterate over any tasks `but` still send this request all the way to `db`. Do we want this behaviour?

Additionally, I don't think the check `maxReadLevel > 0` should have existed in the first place (if we do not allow for empty tasks) since we are guaranteeing that each newer task would have taskID which is increasing in nature, and `maxReadLevel` would be > 0 effectively.

tldr; 

1. Do we still want to have functionality to send in requests to `db` when we have 0 tasks in the array? If yes, we have to add in a check to see how `maxReadLevel` gets updated inside of db.

Comment (ShahabT) [reply service/matching/db.go:223]: right, I think it's fine two skip the write if the list is empty. but that needs to be added at the beginning of CreateTasks and return from the method, right?

File: service/matching/task_writer.go (hunk 2 above)
Comment (ShahabT) [reply service/matching/task_writer.go:233]: the comment is about the order of updating maxReadLevel vs writing tasks. that logic does not live here anymore, so the comment should move with it.

File: service/matching/db.go (hunk 4 above)
Comment (Shivs11) [reply service/matching/db.go:223]: I agree; shall add it thanks

File: service/matching/db.go [hunk 5]
```diff
@@ -231,6 +234,9 @@ func (db *taskQueueDB) CreateTasks(
 			Tasks: tasks,
 		})
 
+	// Update the maxReadLevel after the writes are completed, but before we send the response,
+	// so that taskReader is guaranteed to see the new read level when SpoolTask wakes it up.
+	db.maxReadLevel = maxReadLevel
```
Comment (Shivs11) [inline service/matching/db.go:239]: Moved the assignment after `CreatingTasks` just to be consistent with the old code; 

even here, even if the request to `db` fails (for whatever reason), maxReadLevel is updated and that behaviour is kept consistent with ID allocation and maxReadLevel updates of before.

Comment (ShahabT) [reply service/matching/db.go:239]: thinking more about this comment, it's not relevant anymore because GetMaxReadLevel is locked too. let's get rid of it

File: service/matching/db.go [hunk 6]
```diff
@@ -90,6 +92,20 @@ func (db *taskQueueDB) RangeID() int64 {
 	return db.rangeID
 }
 
+// GetMaxReadLevel returns the current maxReadLevel
+func (db *taskQueueDB) GetMaxReadLevel() int64 {
+	db.RLock()
+	defer db.RUnlock()
+	return db.maxReadLevel
```
[the remark this answers is not in the document]
Comment (Shivs11) [reply service/matching/db.go:99]: Not a Go expert by any means, but if we do use the same locks (which we are in this case since we are using a RW mutex) for this resource, synchronization of values are also taken care of right?  thought this was a given with using the same locking object but yeah curious to see what @dnr states

[the remark this answers is not in the document]
Comment (dnr) [reply service/matching/db.go:99]: Yes, this is fine. I have a little concern that now taskReader will get blocked calling GetMaxReadLevel while we're in the db calling db.store.CreateTasks. In practice it may not be a big deal, but it is a regression.

We could avoid it by keeping maxReadLevel as an atomic, having GetMaxReadLevel only load the atomic without a lock, and SetMaxReadLevel get the lock and then write the atomic.

Effectively that's the same as having two mutexes, one just for maxReadLevel and one for everything else, and always taking the "everything else" before the "maxReadLevel"

[the remark this answers is not in the document]
Comment (Shivs11) [reply service/matching/db.go:99]: Interesting idea @dnr  - seem to be in agreement since it could speed up things with an atomic. Carrying on with the implementation.

[the remark this answers is not in the document]
Comment (Shivs11) [reply service/matching/db.go:99]: on second thought, could you elaborate a little more on this?

Effectively that's the same as having two mutexes, one just for maxReadLevel and one for everything else, and always taking the "everything else" before the "maxReadLevel"

If we do go ahead with this, if `taskReader` calls `GetMaxReadLevel` while `db` is locked and before updating `maxReadLevel`, there is a chance that it could return an older value to it right? Are we okay with this and bank on the fact that `taskReader` will again call `getTaskBatch` and then read the resource again, where it shall have the latest value ?

[the remark this answers is not in the document]
Comment (dnr) [reply service/matching/db.go:99]: If another goroutine has the write lock, GetMaxReadLevel doesn't return an old value, it just blocks. If it were just an in-memory operation that's might be fine, but in this case the write lock is held across a call to the db.

We do want to hold that lock across the db call as a way to serialize db operations (necessary on cassandra at least). So we can't really shrink that lock. But what we could do is put just maxReadLevel under a different lock so we can read it without that large db lock. That's basically how it was before.. you can think of an atomic as just a single value plus a lock just for that value.

[the remark this answers is not in the document]
Comment (ShahabT) [reply service/matching/db.go:99]: Yeah I think we should be fine doing what @dnr suggests. I also did worry a little about locking read path. but did not bring it up thinking we read batches so reads are infrequent and wouldn't be highly impacted. But I don't think that's a valid assumption in may cases where a small backlog is formed for a little while because the poller came a few milliseconds late, and maybe one or two tasks are read from DB every time.

Nevertheless, I think what David is suggesting is good because we don't have to block GetMaxReadLevel to solve the [original problem](https://github.com/temporalio/temporal/pull/5593/files/ecbbac9dd6e62433aaec4069438d65e5c78148ec#r1552145429) which is only about synchronizing UpdateState and CreateTasks.

File: service/matching/db.go (hunk 5 above)
Comment (ShahabT) [reply service/matching/db.go:239]: not that we plan to not lock for GetMaxReadLevel, the order matters, and the comment should be kept.

File: service/matching/db.go [hunk 7]
```diff
@@ -49,11 +50,12 @@ const (
...
-		queue    *PhysicalTaskQueueKey
-		rangeID  int64
-		ackLevel int64
-		store    persistence.TaskManager
-		logger   log.Logger
+		queue        *PhysicalTaskQueueKey
+		rangeID      int64
+		ackLevel     int64
+		store        persistence.TaskManager
+		maxReadLevel int64
```
Comment (dnr) [inline service/matching/db.go:57]: use https://pkg.go.dev/sync/atomic#Int64 instead of a plain value, it's easier to work with and prevents bugs

File: service/matching/db.go [hunk 8]
```diff
@@ -201,6 +232,9 @@ func (db *taskQueueDB) CreateTasks(
 			},
 			Tasks: tasks,
 		})
+
+	atomic.StoreInt64(&db.maxReadLevel, maxReadLevel)
```
Comment (dnr) [inline service/matching/db.go:236]: I thought we were putting the comment back here?

File: service/matching/db.go (hunk 7 above)
Comment (Shivs11) [reply service/matching/db.go:57]: acknowledged

File: service/matching/db.go (hunk 8 above)
Comment (Shivs11) [reply service/matching/db.go:236]: think my github had glitched since I thought Shahab had asked me to put it back but couldn't find it again, so I removed it; shall add it back, sorry for the inconvenience.

File: service/matching/db.go [hunk 9]
```diff
@@ -49,11 +50,12 @@ const (
...
-		queue    *PhysicalTaskQueueKey
-		rangeID  int64
-		ackLevel int64
-		store    persistence.TaskManager
-		logger   log.Logger
+		queue        *PhysicalTaskQueueKey
+		rangeID      int64
+		ackLevel     int64
+		store        persistence.TaskManager
+		maxReadLevel atomic.Int64
```
Comment (dnr) [inline service/matching/db.go:57]:
```suggestion
		maxReadLevel atomic.Int64 // note that even though this is an atomic, it should only be written to while holding the db lock
```

**Review by dnr — APPROVED** [review]
looks good. just one comment suggestion

File: service/matching/db.go (hunk 9 above)
Comment (Shivs11) [reply service/matching/db.go]: For sure - added in.

**Review by ShahabT — APPROVED** [review]
